[Ryuse] iP

Ryuse opened the pull request:
# ✨ Hokmah ✨
> *Chokmah (Hebrew: חָכְמָה, romanized: ḥoḵmā, also transliterated as chokma, chokhmah or hokhma) is the Biblical Hebrew word rendered as "wisdom" in English Bible versions (LXX σοφία sophia, Vulgate sapientia) - [Wikipedia](https://en.wikipedia.org/wiki/Chokmah)

**Hokmah** is a chatbot that doesn't like you but will:
- Organize your tasks for you.
- Run on your computer.
That's it
 

---
## What can it do?
- [x] Add tasks (todos, events, deadlines)
- [x] Delete tasks
- [x] List tasks
- [X] Mark tasks as done
- [X] Find tasks by keyword
- [X] Find tasks by deadline (for deadlines and events)
- [ ] ~~Achieve Sentience~~

- [ ] ~~World Domination~~
---
## Setting up in Intellij
Prerequisites: JDK 17, update Intellij to the most recent version.

1. Open Intellij (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project first)
2. Open the project into Intellij as follows:
   1. Click `Open`.
   2. Select the project directory, and click `OK`.
   3. If there are any further prompts, accept the defaults.
3. Configure the project to use **JDK 17** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
   In the same dialog, set the **Project language level** field to the `SDK default` option.
4. After that, locate the `src/main/java/hokmah.Hokmah.java` file, right-click it, and choose `Run hokmah.Hokmah.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
   ```
   I'm
   ,--.  ,--.         ,--.                         ,--.     
   |  '--'  |  ,---.  |  |,-.  ,--,--,--.  ,--,--. |  ,---.
   |  .--.  | | .-. | |     /  |        | ' ,-.  | |  .-.  |
   |  |  |  | ' '-' ' |  \  \  |  |  |  | \ '-'  | |  | |  |
   `--'  `--'  `---'  `--'`--' `--`--`--'  `--`--' `--' `--'
   
   What do you want?
   ```

---
## For Java Programmers that want to use Hokmah as a library
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Hokmah hokmah = new Hokmah("data/tasks.txt");
        hokmah.run();
    }
}
```

---
### Warning
Keep the `src\main\java` folder as the root folder for Java files (i.e., don't rename those folders or move Java files to another folder outside of this folder path), as this is the default location some tools (e.g., Gradle) expect to find Java files.

Also, Hokmah might bite.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Ryuse committed:
Level-0 and Level-1

- Duke renamed to Hokmah
- gitignore to ignore ".class" files
- README updated
- greet, message, and exit functions added

Ryuse committed:
Level 2

- inputHandler added to handle message types.
- message logging and listing feature is added.

Ryuse committed:
Level-3 to Level-6

- Added ToDo, Deadline, Event, and HokmahException Classes
- Added mark, unmark, delete, todo, deadline, event, and help chatbot commands
- Added Exception Handling
- Added UI Tests

Ryuse committed:
Update README.md

Ryuse committed:
Level-7

- Add SaveHandler to handle saving and loading of tasks from file
- Add saving and loading task from file feature

Ryuse committed:
Merge pull request #1 from Ryuse/branch-Level-7

Level-7

Ryuse committed:
Add datetime support for tasks

- add datetime support for tasks for better date formatting and understanding

Ryuse committed:
Merge pull request #2 from Ryuse/branch-Level-8

Add datetime support for tasks

Ryuse committed:
refactored code into different classes for easy code management

- Split Hokmah class into UiHandler, TaskList, InputHandler, and CommandHandler to make code cleaner and easier to manager in the future
- New Ui changes added for better visuals when Hokmah is responding

Ryuse committed:
Merge pull request #3 from Ryuse/branch-A-MoreOOP

refactored code into different classes for easy code management

Ryuse committed:
move java files to packages

Ryuse committed:
Merge pull request #4 from Ryuse/branch-A-Packages

move java files to packages

Ryuse committed:
Merge pull request #5 from Ryuse/add-gradle-support

Add gradle support

Ryuse committed:
Update build.gradle

- Update build.gradle to use Hokmah as main class

Ryuse committed:
Merge pull request #6 from Ryuse/branch-A-Gradle

Update build.gradle

Ryuse committed:
Add unit test codes and bugfixes

- Add test codes to test add task, save, load, and print commands
- Fixed getting task via index

Ryuse committed:
Merge pull request #7 from Ryuse/branch-A-JUnit

Add unit test codes and bugfixes

Ryuse committed:
Update build.gradle

- Update build.gradle to change jar name to hokmah

Ryuse committed:
Merge pull request #8 from Ryuse/branch-A-Jar

Update build.gradle

Ryuse committed:
Add javadocs headers to non-private methods

Ryuse committed:
Fixed indentation

Ryuse committed:
Add header docs to classes

Ryuse committed:
Add find command feature

- Add find command feature to filter list of tasks based on keyword provided

Ryuse committed:
Merge pull request #9 from Ryuse/branch-A-JavaDoc

Branch a java doc

Ryuse committed:
Merge branch 'master' into branch-A-CodingStandard

Ryuse committed:
Merge pull request #10 from Ryuse/branch-A-CodingStandard

Fixed indentation

Ryuse committed:
Merge branch 'master' into branch-Level-9

Ryuse committed:
Merge pull request #11 from Ryuse/branch-Level-9

Add find command feature

Ryuse committed:
Merge bugfixes and indentation adjustment

Ryuse committed:
Edit README.md

- Edit README.md to show more details about the project

Ryuse committed:
Command Handler and UI Handler Bugfix

- Bugfix for upcoming command
- Command prints moved to UI
- Personality added for responses

Ryuse committed:
Merge pull request #12 from Ryuse/branch-PersonalityAndReadMe

 Personality and README changes

Ryuse committed:
Add Checkstyle to Gradle

- Add Checkstyle to Gradle for code format checking

Ryuse committed:
Reactor code based on Checkstyle

Ryuse committed:
Merge pull request #13 from Ryuse/branch-A-CheckStyle

Branch a check style

Ryuse committed:
Add JavaFX to project

- Add JavaFX to project to show UI
- Refactored code to suit JavaFX

Ryuse committed:
Merge pull request #14 from Ryuse/branch-Level-10

Add JavaFX to project

Ryuse committed:
Add default filepath to initialize Hokmah

- Add default filepath to initialize Hokmah to allow empty arguments

bkkuan commented on src/main/java/Task.java:
```diff
@@ -17,17 +14,25 @@ public void unmarkDone() {
-//    public int getId() {
-//        return id;
-//    }
+    public String getTimeStart() {
```
Poor method placement. getTimeStart return null in Task. Task should be abstract or removed

bkkuan commented on src/main/java/Task.java:
```diff
@@ -1,12 +1,9 @@
 public class Task {
     private boolean isDone = false;
-//    private static int lastId = 0;
```
Commented out code should be removed

bkkuan commented on src/main/java/Task.java:
```diff
@@ -17,17 +14,25 @@ public void unmarkDone() {
         isDone = false;
     }
 
-    public boolean getDone() {
```
this method appears to start but never properly closes before the next method getType() begins.
Missing closing brace or unintended method nesting

bkkuan commented on src/main/java/Task.java:
```diff
@@ -17,17 +14,25 @@ public void unmarkDone() {
+        return null;
+    }
 
     public String toString() {
```
Override annotation is missing when overriding a method

bkkuan reviewed (commented):
The separation of concerns among different task types and the integration of file-saving/loading functionality is well executed. Keep up the excellent work!

datletu commented on src/main/java/hokmah/task/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Constructs an empty TaskList.
+     */
+    public TaskList() {
```
Not sure if this is needed for your program workflow but would it be better if you add a TaskList(ArrayList<Task> tasks) that allows you to make an arraylist your task list (not just create an empty list), just as an additional option?

datletu commented on src/main/java/hokmah/exception/HokmahException.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Enumeration of possible exception types.
+     */
+    public enum ExceptionType {
```
Cool use of enum

datletu commented on src/test/java/hokmah/HokmahTest.java:
```diff
@@ -0,0 +1,126 @@
+package hokmah;
```
Nice robust testing

datletu commented on src/main/java/hokmah/command/MessageHandler.java:
```diff
@@ -0,0 +1,211 @@
+package hokmah.command;
```
I like the attitude of the output lol

datletu reviewed (approved):
Overall nicely done I didn't see anything egregious violation of the coding standard

Ryuse committed:
Command Package: Add Assertions

Class files in Command package are not asserted making
them functions in the class files more prone to errors.

Add assertion to functions in command package to assert
functions' behaviour is working as intended.

Assertion has been added to the start of the functions to
assert that the inputs are not null and are of expected values.

Ryuse committed:
view.Main.java: Add assertions

Add assertion to the minimum height  and width of the screen.

This is to make sure the height and width of the screen is of
the right size when the app is run.

Ryuse committed:
Update MainWindow.java

Hokmah doesn't show response before exiting the app through
command.

Now, exiting the app through command now shows response
before the app closes.

Ryuse committed:
Command package: refactor classes

MessageHandler and CommandHandler has been difficult to read and
understand the flow. This may cause debugging to be difficult in the
future.

String formatting and multiline string has been used in MessageHandler
for the messages and checks for tasks and parsing has been isolated
for better read of flow, making the happy cases more prominent.

These changes improve code maintainability through consistent formatting
patterns while ensuring accurate task display in filtered views.

Ryuse committed:
Refactor MessageHandler.java to use Streams

The `getFindMessage` and `getUpcomingTasksOnMessage` methods in `MessageHandler.java` currently use loops to process and format task lists.
This approach can be verbose and less readable.

Using Streams improves code readability and conciseness by leveraging functional programming constructs.

Let's refactor these methods to use Streams:
- Replace the loop in `getFindMessage` with a stream to map and join matching tasks.
- Replace the loop in `getUpcomingTasksOnMessage` with a stream to format and join upcoming tasks.

This change aligns with Java practices and reduces boilerplate code while maintaining the same functionality. The refactored code is more maintainable and easier to extend in the future.

Ryuse committed:
Merge pull request #15 from Ryuse/branch-A-Assertions

Add assertion to Hokmah command and view packages

Ryuse committed:
Merge branch 'master' into branch-A-CodeQuality

Ryuse committed:
Merge pull request #16 from Ryuse/branch-A-CodeQuality

Merge CodeQuality branch to master branch

Ryuse committed:
Merge branch 'master' into branch-A-Streams

Ryuse committed:
Refactor CommandHandler and MessageHandler

Merging master to A-Streams did not merge the Assert Branch properly due
to a previous marge mistake. This commit aims to add the missing assert
statements from the assert branch.

The change is done to add the assert statements and help make the code
more bug free by asserting possible points of the code that can give bugs.

What's done about it is that the assert statements from the Assert Branch has been
added back in to the `CommandHandler.java` and `MessageHandler.java`
by trying to reverse the wrong merge commit.

As for why it's done this way, the reversal of commit helps with
reducing redundant work and just let software like GitHub Desktop
handle changes by clicking on changes to be kept

Ryuse committed:
Update CommandHandler.java

Ryuse committed:
Merge pull request #17 from Ryuse/branch-A-Streams

Merge branch A-Stream into master

Ryuse committed:
Create gradle.yml

Create a `gradle.yml` file for CI workflows to check codes in push or pull requests.

Ryuse committed:
Update `runtest.bat` and `runtest.sh`

Ryuse committed:
Update gradle.yml

Ryuse committed:
Merge pull request #18 from Ryuse/branch-A-CI

Create gradle.yml

Ryuse committed:
Add window for help command

Help command had no window before to show all of the commands leading to less readable
UI.

Using JavaFX with CSS to create a new window for the help command increases the
readability of the list of the commands while allowing the window of list of commands
to be kept to the side for reference.

Let's
- Remove the old list of commands in the `MessageHandler.java` and transform it
into a table in the help window
- Rename the window Java files into Handlers to handle the windows
- Create a new help window with respective CSS, FXML, and Java files

Ryuse committed:
Refactor files for checkstyle

Ryuse committed:
Merge pull request #19 from Ryuse/branch-C-Help

Add window for help command

Ryuse committed:
Refactor SaveHandler for readability

- `SaveHandler.java` methods extracted out for better readability
- New Exception message added for no file found when running the program
- New response when listing task with no task

Ryuse committed:
Add multiline messaging UI feature

Messages sent by Hokmah used to be only one full dialog box for
the UI.

With this commit, Hokmah now has the ability to send messages
through multiple dialog box by making the MessageHandler return a
list of message lines instead of a full string.

Why it was done this way was to add more customizability and flavour
when showing messages through the UI.

Let's
- Change the `hokmah.command` and `hokmah.exception` packages
to give responses as array of String rather than a full String
- Change `MainWindowController.java` to show responses as multiple
dialog boxes instead of just one full one.

Ryuse committed:
Change Hokmah UI elements

- Add icons and background images for the UI
- Change window resolution to suit the feel of a messaging app

Ryuse committed:
Update CommandHandler.java

Ryuse committed:
Merge pull request #21 from Ryuse/branch-A-BetterGui

Merge Branch A-BetterGui

Ryuse committed:
Create Hokmah User Guide

Ryuse committed:
Refactor `hokmah` package for error handling

Ryuse committed:
Add welcome message to UI

Ryuse committed:
Change help and exception messages

Ryuse committed:
Rename UI image for user guide

Ryuse committed:
Fix exception checkstyle

Ryuse committed:
Update user guide UI and message

Ryuse committed:
Bug fix: trim inputs for accidental spaces

Ryuse committed:
Bug fix: help command window popup

`help` command does not show windows when closing
via the windows close button.

To fix this, an event listener has been added which
sets the tracker of the help window visibility to the right
variable.

Ryuse committed:
Refactor `InputHandler.java` for checkstyle

Ryuse committed:
Change `upcoming` command to `upcomingOn`

Ryuse committed:
Refactor `hokmah` package for code quality

- Refactored `CommandHandler.java` for
better code quality
- Changed `HokmahException.java` exception
messages for better understanding

Ryuse committed:
Update User Guide

Ryuse committed:
Add unit tests for `CommandHandler.java`

Ryuse committed:
Update User Guide for running command

Ryuse committed:
Bugfix: `mark` and `unmark` Commands

`mark and `unmark` commands used to not save
the task list state when run by themselves.

This commit makes it so that the state of the
task list is saved after the command call.

Parts of `CommandHandler.java` has also
been refactored for better code readability.
